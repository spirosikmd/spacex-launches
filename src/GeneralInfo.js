import React from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import { LaunchPropType } from './Launch';

export function getNumberOfSuccessful(launches) {
  return launches.filter(launch => launch.isSuccessful).length;
}

export function getNumberOfUpcoming(launches) {
  return launches.filter(launch => launch.isUpcoming).length;
}

export function getNumberOfFailed(launches) {
  return launches.filter(launch => launch.isFailed).length;
}

export const GeneralInfo = ({ launches }) => {
  return (
    <>
      <Typography>Total: {launches.length}</Typography>
      <Typography>Successful: {getNumberOfSuccessful(launches)}</Typography>
      <Typography>Failed: {getNumberOfFailed(launches)}</Typography>
      <Typography>Upcoming: {getNumberOfUpcoming(launches)}</Typography>
    </>
  );
};

GeneralInfo.propTypes = {
  launches: PropTypes.arrayOf(PropTypes.shape(LaunchPropType)).isRequired,
};

export default React.memo(GeneralInfo);
