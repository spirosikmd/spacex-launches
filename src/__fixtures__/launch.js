export function createLaunchResponse(data) {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data) {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
