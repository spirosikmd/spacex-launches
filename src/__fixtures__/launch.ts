import { LaunchResponseData, LaunchData } from '../api';

export function createLaunchResponse(
  data: LaunchResponseData
): LaunchResponseData {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data: LaunchData): LaunchData {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
