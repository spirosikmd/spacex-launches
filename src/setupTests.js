import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createSerializer } from 'enzyme-to-json';
import { unwrap } from '@material-ui/core/test-utils';

configure({ adapter: new Adapter() });

expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));

global.shallow = (Component, props = {}) => {
  const ComponentWithoutStyles = unwrap(Component);
  return shallow(<ComponentWithoutStyles {...props} />);
};

global.mount = (Component, props = {}) => {
  const ComponentWithoutStyles = unwrap(Component);
  return mount(<ComponentWithoutStyles {...props} />);
};

global.Date = jest.fn().mockImplementation(() => {
  return {
    toUTCString: jest.fn().mockReturnValue('Fri, 11 Jan 2019 15:31:00 GMT'),
    getDate: jest.fn().mockReturnValue(3),
    getMonth: jest.fn().mockReturnValue(10),
    getFullYear: jest.fn().mockReturnValue(2019),
  };
});

// https://github.com/airbnb/enzyme/issues/1875
jest.mock('react', () => {
  const r = jest.requireActual('react');
  return { ...r, memo: x => x };
});
