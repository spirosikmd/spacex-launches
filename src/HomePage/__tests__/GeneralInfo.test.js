import React from 'react';
import { render } from '../../setupTests';
import GeneralInfo from '../GeneralInfo';
import { createLaunch } from '../../__fixtures__/launch';

describe('GeneralInfo', () => {
  let props;

  beforeEach(() => {
    props = {
      launches: [
        createLaunch(),
        createLaunch({ isSuccessful: false, isFailed: true }),
        createLaunch({ isSuccessful: false, isUpcoming: true }),
      ],
    };
  });

  it('renders general info about launches', () => {
    const { asFragment } = render(<GeneralInfo {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
