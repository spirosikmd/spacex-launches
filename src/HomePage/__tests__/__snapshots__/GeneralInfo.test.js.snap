// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneralInfo renders general info about launches 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    Total: 3
  </p>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    Successful: 1
  </p>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    Failed: 1
  </p>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    Upcoming: 1
  </p>
</DocumentFragment>
`;
